src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -0,0 +1,333 @@
+}
+
+absl::Status WSAErrorToStatusWithMessage(
+    int wsa_error, absl::string_view custom_message,
+    const grpc_core::DebugLocation& location) {
+  // See
+  // https://learn.microsoft.com/en-us/windows/win32/api/Winsock2/nf-winsock2-wsasend
+  // https://learn.microsoft.com/en-us/windows/win32/api/Winsock2/nf-winsock2-wsarecv
+  char* wsa_message = gpr_format_message(wsa_error);
+  std::string message;
+  if (!custom_message.empty()) {
+    std::string message = absl::StrCat(wsa_message, ": ", custom_message);
+  } else {
+    std::string message = wsa_message;
+  }
+  gpr_free(wsa_message);
+  switch (wsa_error) {
+    case 0:
+      return absl::OkStatus();
+    case WSAECONNRESET:
+    case WSAECONNABORTED:
+    case WSA_OPERATION_ABORTED:
+      return StatusCreate(absl::StatusCode::kAborted, message, location, {});
+    case WSAETIMEDOUT:
+      return StatusCreate(absl::StatusCode::kDeadlineExceeded, message,
+                          location, {});
+    case WSAEFAULT:
+    case WSAEINVAL:
+      return StatusCreate(absl::StatusCode::kInvalidArgument, message, location,
+                          {});
+    case WSAENETDOWN:
+    case WSAENOTCONN:
+    case WSAESHUTDOWN:
+      return StatusCreate(absl::StatusCode::kUnavailable, message, location,
+                          {});
+    default:
+      return StatusCreate(absl::StatusCode::kUnknown, message, location, {});
+  }
+}
+
+absl::Status WSAErrorToStatus(int wsa_error,
+                              const grpc_core::DebugLocation& location) {
+  return WSAErrorToStatusWithMessage(wsa_error, "", location);
+}
+
+}  // namespace
+
+WindowsEndpoint::WindowsEndpoint(
+    const EventEngine::ResolvedAddress& peer_address,
+    std::unique_ptr<WinSocket> socket, MemoryAllocator&& allocator,
+    const EndpointConfig& /* config */, Executor* executor)
+    : peer_address_(peer_address),
+      socket_(std::move(socket)),
+      allocator_(std::move(allocator)),
+      handle_read_event_(this),
+      handle_write_event_(this),
+      executor_(executor) {
+  sockaddr addr;
+  int addr_len = sizeof(addr);
+  if (getsockname(socket_->socket(), &addr, &addr_len) < 0) {
+    GPR_ASSERT(false &&
+               "Unrecoverable error: Failed to get local socket name.");
+  }
+  local_address_ = EventEngine::ResolvedAddress(&addr, addr_len);
+  local_address_string_ = *ResolvedAddressToURI(local_address_);
+  peer_address_string_ = *ResolvedAddressToURI(peer_address_);
+}
+
+WindowsEndpoint::~WindowsEndpoint() {
+  socket_->MaybeShutdown(absl::OkStatus());
+}
+
+void WindowsEndpoint::Read(absl::AnyInvocable<void(absl::Status)> on_read,
+                           SliceBuffer* buffer, const ReadArgs* /* args */) {
+  // TODO(hork): last_read_buffer from iomgr: Is it only garbage, or optimized?
+  GRPC_EVENT_ENGINE_TRACE("WindowsEndpoint::%p reading", this);
+  // Prepare the WSABUF struct
+  WSABUF wsa_buffers[kMaxWSABUFCount];
+  // TODO(hork): use read hint instead of the default?
```

Comment:
ideally we should not return from a read until the read hint is achieved (it's labelled `min_progress_size` in iomgr and that seems a better name)

if we do so, the calling code will assuredly call read again and not process the incoming memory until that limit is achieved anyway

whether that's important now or not I have no opinion, but ultimately we'll want to post *at least* min_progress_size bytes for an async read to avoid spurious wakeups from the kernel (we could post more, that would be fine)

Reply:
Agreed. `min_progress_size` is currently unused in the Windows iomgr implementation, but we can improve upon that here. I've added the override.